### PR TITLE
fixes #18235 - proc annotation type macro sym leak

### LINKED ISSUE #18235: The original AST of an proc modified by typed macros still persist across module boundary
### Example
`a.nim`:
```nim
import macros

macro rename(n: typed): untyped =
  result = copyNimTree(n)
  # turn exported nnkSym -> nnkPostfix(*, nnkIdent), forcing re-sem
  result[0] = nnkPostfix.newTree(ident"*").add:
    n.name.strVal.ident

macro unexport(n: typed): untyped =
  result = copyNimTree(n)
  # turn nnkSym -> nnkIdent, forcing re-sem and dropping any exported-ness
  # that might be present
  result[0] = n.name.strVal.ident

proc foo*() {.unexport.} = discard
```

`b.nim` (compile this file):
```nim
import a
foo()
```

### Current Output
Compilation successful

### Expected Output
Error about `foo` not existing/exported.

### Additional Information
* If you use `rename`, which is what CPS would do (force a re-sem on the bootstrap to take the original place), you will be met with this error:
  ```
  Error: ambiguous call; both a.foo() [proc declared in a.nim(15, 6)] and a.foo() [proc declared in a.nim(7, 18)] match for: ()
  ```

* This is blocking CPS procs from being used across module boundary

```
$ nim -v
Nim Compiler Version 1.5.1 [Linux: amd64]
Compiled at 2021-06-08
Copyright (c) 2006-2021 by Andreas Rumpf

git hash: 51ab7ccec1ffa21cef67f1a75dea889653751cfc
active boot switches: -d:release -d:nimUseLinenoise
```

### PR DESCRIPTION
**Pull request by saem**
- also fixed a typo
- proc annotations guard symbol exports with shadow scopes
- symbol handling is shadow scope aware

fixes #18235

### CODE REVIEW

**Comment by saem** [comment]
@Clyybber and @alaviss assuming the CI passes, here is the fix me thinks.

File: tests/macros/m18235.nim [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  result[0] = nnkPostfix.newTree(ident"*").add:
+    n.name.strVal.ident
+
+macro unexport(n: typed): untyped =
+  result = copyNimTree(n)
+  # turn nnkSym -> nnkIdent, forcing re-sem and dropping any exported-ness
+  # that might be present
+  result[0] = n.name.strVal.ident
+
+proc foo*() {.unexport.} = discard
```
Comment (alaviss) [inline tests/macros/m18235.nim]: You should probably do one more test with `rename` so that we know it also works.

Comment (saem) [reply tests/macros/m18235.nim]: ugh, stop making so much sense. :P

(goes and writes a test)

File: tests/macros/t18235.nim [hunk 2]
```diff
@@ -0,0 +1,10 @@
+discard """
+errormsg: '''
+undeclared identifier: 'foo'
+'''
+"""
+import m18235
```
Comment (timotheecour) [inline tests/macros/t18235.nim:1]: combine this test inside t18235_rename_export.nim and call:

doAssert not declared(foo)

File: tests/macros/m18235.nim [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import macros
+
+# Necessary code to update the AST on a symbol across module boundaries when
+# processed by a type macro. Used by a test of a corresponding name of this
+# file.
+
+macro rename(n: typed): untyped =
```
Comment (timotheecour) [inline tests/macros/m18235.nim]: rename as exportSym?

**Comment by timotheecour** [comment]
@saem please add `fixes #18235` to PR description (see rationale in contributing.html)

File: compiler/semstmts.nim [hunk 4]
```diff
@@ -1916,9 +1918,9 @@ proc semProcAux(c: PContext, n: PNode, kind: TSymKind,
 
   if not hasProto and sfGenSym notin s.flags: #and not isAnon:
     if s.kind in OverloadableSyms:
-      addInterfaceOverloadableSymAt(c, delcarationScope, s)
+      addInterfaceOverloadableSymAt(c, declarationScope, s)
     else:
-      addInterfaceDeclAt(c, delcarationScope, s)
+      addInterfaceDeclAt(c, declarationScope, s)
```
Comment (saem) [inline compiler/semstmts.nim:1921]: The addition to the interface should happen after the potential reset of the scope if a prototype is found, around line 1955.

Comment (saem) [reply compiler/semstmts.nim:1921]: Also document the heck out of why that's important.

File: compiler/lookups.nim [hunk 5]
```diff
@@ -340,22 +340,35 @@ proc addOverloadableSymAt*(c: PContext; scope: PScope, fn: PSym) =
 proc addInterfaceDecl*(c: PContext, sym: PSym) =
   # it adds the symbol to the interface if appropriate
   addDecl(c, sym)
-  addInterfaceDeclAux(c, sym)
+  if not c.currentScope.isShadowScope:
```
Comment (saem) [inline compiler/lookups.nim:349]: These don't cover all the entry points, but it's also not clear what should happen. :(

File: tests/macros/t18235.nim [hunk 6]
```diff
@@ -1,10 +1,18 @@
-discard """
-errormsg: '''
-undeclared identifier: 'foo'
-'''
-"""
 import m18235
 
 # this must error out because it was never actually exported
+doAssert(not declared(foo))
```
Comment (timotheecour) [inline tests/macros/t18235.nim:4]: optional nit, but `doAssert not declared(foo)` is the more common style in tests

File: tests/macros/m18235.nim [hunk 7]
```diff
@@ -4,7 +4,7 @@ import macros
 # processed by a type macro. Used by a test of a corresponding name of this
 # file.
 
-macro rename(n: typed): untyped =
+macro eexport(n: typed): untyped =
```
Comment (timotheecour) [inline tests/macros/m18235.nim:7]: reexport?

File: compiler/lookups.nim [hunk 8]
```diff
@@ -359,7 +363,7 @@ proc openShadowScope*(c: PContext) =
                           depthLevel: c.scopeDepth)
 
 proc closeShadowScope*(c: PContext) =
-  ## closes the shadow scope, but doesn't merge any of the symbols
+  ## closes the shadow scope, but doesn't merge any of the symbols
```
Comment (timotheecour) [inline compiler/lookups.nim]:
```suggestion
  ## closes the shadow scope, but doesn't merge any of the symbols
```

File: tests/macros/m18235.nim [hunk 9]
```diff
@@ -17,4 +17,26 @@ macro unexport(n: typed): untyped =
...
+    result[i] = newCall(ident"eexport", result[i])
+
+macro unexportMulti(n: typed): untyped =
+  # use the call version of `unexport` macro for one or more decls
+  result = copyNimTree(n)
+  for i in 0..<result.len:
+    result[i] = newCall(ident"unexport", result[i])
+
+unexportMulti:
+  proc oof*() = discard
```
Comment (timotheecour) [inline tests/macros/m18235.nim:38]:
```suggestion
  proc oof*() = discard
  proc oof2() = discard
```
(optional; it case it also handles this easily)

Comment (Clyybber) [reply tests/macros/m18235.nim:38]: unexport doesnt handle this

File: tests/macros/m18235.nim (hunk 7 above)
Comment (Clyybber) [reply tests/macros/m18235.nim:7]: "export" makes more sense than "reexport" so lets keep it as is, just a test anyways

File: tests/macros/t18235.nim (hunk 6 above)
Comment (Clyybber) [reply tests/macros/t18235.nim:4]: it cant hurt to test other styles too

**Review by Clyybber — APPROVED** [review]

**Pull request merged**